### PR TITLE
Fixed #14611 -- Added query_params argument to RequestFactory and Client classes.

### PR DESCRIPTION
**Pull request by knyghty**

### CODE REVIEW

File: django/test/client.py [hunk 1]
```diff
@@ -494,26 +507,37 @@ def post(
...
-    def head(self, path, data=None, secure=False, *, headers=None, **extra):
+    def head(
+        self, path, data=None, secure=False, *, headers=None, query_params=None, **extra
+    ):
         """Construct a HEAD request."""
-        data = {} if data is None else data
+        if query_params and data:
+            raise ValueError("You cannot specify query_params and data together.")
+        query_params = query_params or data
+        query_params = {} if query_params is None else query_params
```
Comment (shangxiao) [inline django/test/client.py:521]: Hm so we now have 2 kwargs that do the same thing just named differently… doesn't seem Pythonic 😅  (or Djangonic)

The forum discussion didn't really talk about what to do with get() & whether we have 2 or deprecate data or what.  I wonder whether it's worth asking @adamchainz & co.

Comment (knyghty) [reply django/test/client.py:521]: I was thinking about this as well. I'd be keen to deprecate it if only to make the code a little less messy and more consistent. There is a lot of code using this, but only in tests and I feel like breaking people's tests is tolerable relative to breaking their sites.

On the other hand, the current API has been there a long time and it works, even if it's a little confusing.

File: docs/topics/testing/tools.txt [hunk 2]
```diff
@@ -158,25 +158,28 @@ Use the ``django.test.Client`` class to make requests.
     Once you have a ``Client`` instance, you can call any of the following
     methods:
 
-    .. method:: Client.get(path, data=None, follow=False, secure=False, *, headers=None, **extra)
+    .. method:: Client.get(path, data=None, follow=False, secure=False, *, headers=None, query_params=None, **extra)
```
Comment (felixxm) [inline docs/topics/testing/tools.txt:161]: `versionchanged` annotations are missing for every method.

File: django/test/client.py (hunk 1 above)
Comment (felixxm) [reply django/test/client.py:521]: I agree that we cannot have two parameters for the same. Personally, I wouldn't add `query_params` to `get()` and `head()`. I think we need to reach a wider audience and ask on the forum.

Comment (knyghty) [reply django/test/client.py:521]: I added a post with my rationale: https://forum.djangoproject.com/t/handling-test-client-data-query-params/25409

Comment (carltongibson) [reply django/test/client.py:521]: Do they do the same thing? `query_params` explicitly sets the query string. `data` sets form data on the request in accordance with the HTTP rules for the method in use. (It's an implementation detail that that happens to be on the query string.)

I may want to use the `query_params` **or**  send form `data`, depending on my test. 

I just can't use **both** at the same time. 

A form like this...

```
<form method="get" action="/test-view/?a=1&b=2">
    <input type="hidden" name="c" value="3">
    <input type="hidden" name="d" value="4">
    <input type="submit" value="Submit">
</form>
```

… submits to `/test-view/?c=3&d=4`, so `get()` and `head()` should set `data` in place of `query_params` if both are provided, and raise a warning — or error if you prefer — that `query_params` cannot be used in this case. 

tl;dr: I think having both is fine. It keeps the API consistent. It's **using** both **at the same time** that we should guard against

I'd be :-1: to removing the `data` argument, as a whole ecosystem's worth of tests would need updating, and it's not worth that. (`django-upgrade` is available in theory but people have to learn about it and then use it, which is not a given...)

File: django/test/client.py [hunk 3]
```diff
@@ -494,26 +507,37 @@ def post(
...
 
-    def head(self, path, data=None, secure=False, *, headers=None, **extra):
+    def head(
+        self, path, data=None, secure=False, *, headers=None, query_params=None, **extra
+    ):
         """Construct a HEAD request."""
-        data = {} if data is None else data
+        if query_params and data:
+            raise ValueError("You cannot specify query_params and data together.")
+        query_params = query_params or data
```
Comment (carltongibson) [inline django/test/client.py:520]: This should favour `data`, as per the form example in https://github.com/django/django/pull/17447/files/a261b7669935faa3d27fee2a03ab6f45436a8bd3#r1397428841

File: django/test/client.py (hunk 1 above)
Comment (shangxiao) [reply django/test/client.py:521]: Yeah but following that logic `query_params` would never get sent 🤔 because:

```
<form method="get" action="/test-view/?a=1&b=2">
    <input type="submit" value="Submit">
</form>
```

submits to `/test-view/?`

ie query params are always ignored.

which is even better argument for just having `data` 😬

Comment (carltongibson) [reply django/test/client.py:521]: If you set `data` that's a _submitting a form_ type request, so **Don't use `query_params` here** is (would be) great!

But there's **loads** of times when I'm making `GET` requests that aren't form submissions and `query_params` is the natural argument in that case. 

(Or so it seems to me 😜)

**Comment by felixxm** [comment]
@knyghty Thanks :+1: We can keep both `data` and `query_params` for `GET` and `HEAD`. Do you have time to keep working on remaining comments (https://github.com/django/django/pull/17447#discussion_r1397443016, https://github.com/django/django/pull/17447#discussion_r1395342985)?

**Comment by knyghty** [comment]
@felixxm Yes, this week I think, thanks 👍🏻

**Comment by knyghty** [comment]
@felixxm these are both done now.

File: django/test/client.py [hunk 4]
```diff
@@ -459,18 +468,21 @@ def _get_path(self, parsed):
...
         return path.decode("iso-8859-1")
 
-    def get(self, path, data=None, secure=False, *, headers=None, **extra):
+    def get(
+        self, path, data=None, secure=False, *, headers=None, query_params=None, **extra
+    ):
         """Construct a GET request."""
-        data = {} if data is None else data
+        if query_params and data:
+            raise ValueError("You cannot specify query_params and data together.")
```
Comment (carltongibson) [inline django/test/client.py:476]: Was this tested?

Comment (knyghty) [reply django/test/client.py:476]: Ok, you caught me. I added tests for this. Do you think we also need tests to make sure you _can_ use both for post/put/patch, or is that unnecessary?

Comment (carltongibson) [reply django/test/client.py:476]: The exceptional behaviour seems worth testing. 

A POST with `data` to a URL with `query_params` might be interesting for documentation/regression purposes. (But it's not something I'd insist on: what's your thought?)

Comment (knyghty) [reply django/test/client.py:476]: Considering that's a primary use case of the PR, seems like a good idea. Added tests around this and some others I missed.

File: docs/topics/testing/tools.txt (hunk 2 above)
Comment (felixxm) [reply docs/topics/testing/tools.txt:161]: `query_params=None` is missing in `AsyncClient` and `Client` signatures and docs.

File: docs/topics/testing/tools.txt [hunk 5]
```diff
@@ -345,12 +358,20 @@ Use the ``django.test.Client`` class to make requests.
 
     .. method:: Client.head(path, data=None, follow=False, secure=False, *, headers=None, **extra)
 
+        .. versionchanged:: 5.1
+
+            The ``query_params`` argument was added.
+
```
Comment (felixxm) [inline docs/topics/testing/tools.txt:364]: `query_params` is missing in the `head()` signature and docs.

**Review by felixxm — APPROVED** [review]
@knyghty Thanks :+1: I pushed missing documentation changes.
